### ISSUE TITLE
Random numbers in solutions

### ISSUE THREAD

**Issue by Haydnspass** [DerThorsten/mlcv#1]
Hi!
when it comes to solutions with random numbers I thought it would be cool to have them with fixed (and "public" seed) so it's easy for us to compare against the true solution (urs) and among our colleagues.
I thought of something like

```python
import random
random.seed(0)
```
so one draws the same but 'random' numbers every time.
That would be cool for assignment 6 especially.

**Comment by raphaelm** [comment]
Unfortunately impractical, since the RNGs ``random`` and ``numpy.random`` apparently behave to differently.

**Comment by DerThorsten** [comment]
I guess :
    
    import random
    import numpy

    random.seed(42)
    numpy.random.seed(42)

should work (UNTESTED!)
But this is will not lead to the same random numbers on different machines / python version/ etc

**Comment by raphaelm** [comment]
In that case, random.random and numpy.random will generate different numbers and yes, it might differ between machines

**Comment by Haydnspass** [comment]
Hm, I do not get different results on my two machines nor do I get any difference between python 2.7 and anaconda 3.6, however I do get different results for np and random.

**Comment by DerThorsten** [comment]
But no standard will guarantee that!
In case you don't trust my word rtfm.

Furthermore it still depends on the order you call your random number generator etc.
And last but not least...i do not really care for this comparability =)

**Comment by raphaelm** [comment]
> In case you don't trust my word rtfm.

Actually, Python's [``random.seed``](https://docs.python.org/3/library/random.html?highlight=random#random.seed) function **has** a consistency/compatibility guarantee, that's why it takes a second parameter ``version``. For example, in 3.2 the behaviour changed if you call ``random.seed`` with a non-integer, but you can force the old behaviour by setting ``version=1``. Similarly, we could set ``version=2`` and be sure that the random numbers kee
p today's behaviour. This guarantee is given explicitly [here](https://docs.python.org/3/library/random.html?highlight=random#notes-on-reproducibility). Other implementations like PyPy implement the same Mersenne-Twister as CPython and are therefore also compatible. 😉

**Comment by DerThorsten** [comment]
@raphaelm Am i missing something or why would this be a guarantee for consistency across different machines?

**Comment by raphaelm** [comment]
"By re-using a seed value, the same sequence should be reproducible from run to run as long as multiple threads are not running." sounds like a pretty nonsensical statement to me if it would only hold true on one machine. However, yes, while its not explicitly state, there's also no reason why it would work differently on different machines since the algorithm is well-defined and only depends on the seed.

**Comment by DerThorsten** [comment]
This is just a guarantee on one machine

**Comment by raphaelm** [comment]
I think the docs is unclear at that point and both of our interpretations are valid if only looking at the words, but in practice it *is* consistent across machines and would be pretty useless otherwise.

numpy has an [even stronger guarantee](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.RandomState.html#numpy.random.RandomState), by the way.

**Comment by DerThorsten** [comment]
If you compile your own python, and compile it with different flags.
Or even worse, if your are on an very old cpu which does not support some operations.
Why would the floating point numbers be the same....
Even non randomized algorithms tend to give different results on different machines

**Comment by DerThorsten** [comment]
@raphaelm    numpy:  "up to roundoff error " ... so this means what...in practice there are no real guarantees when it comes to floating points..

**Comment by raphaelm** [comment]
> in practice there are no real guarantees when it comes to floating points..

You are right, point taken. However, from what I understand the Mersenne-Twister itself only consists of integer operations, so floating points only come into play in the ``random.random`` call itself and therefore can't accumulate. So yes, while roundoff errors might be inconsistent between machines, the given guarantee is probably more than good enough for our purpose.
